### ISSUE TITLE
Keep alive

### ISSUE THREAD

**Issue by ostcar** [OpenSlides/openslides3-autoupdate-service#19]
From #16 
>  Chromium kills the connection after some time (in both cases, I didn't get any headers or so):
>       
>       * If there was never a message (since the client is up to date) after 240s the connection gets closed:
> 
> 
> ```
> net::ERR_EMPTY_RESPONSE
> 
> status: 0
> statusText: "Unknown Error"
> url: "/system/autoupdate"
> ok: false
> name: "HttpErrorResponse"
> message: "Http failure response for /system/autoupdate: 0 Unknown Error"
> ```
> 
>     * If there was a message, after 120s the connection is closed, if no new message arrived:
> 
> 
> ```
> net::ERR_INCOMPLETE_CHUNKED_ENCODING 200 (OK)
> 
> status: 0
> statusText: "Unknown Error"
> url: "http://localhost:4200/system/autoupdate?change_id=135252742000"
> ok: false
> name: "HttpErrorResponse"
> message: "Http failure response for http://localhost:4200/system/autoupdate?change_id=135252742000: 0 Unknown Error"
> ```
> 
> I'm not sure about other browsers, but I can see the reason: Since this is still a "normal" http request, it is closed, if there is no activity. See: https://tools.ietf.org/id/draft-loreto-http-bidirectional-07.html
> -> We have two possibilities:
> 
>     1. Send a layer-7 keep alive. E.g. a message with the key `keep-alive`, which gets discarded in the client code.
> 
>     2. The client always must reopen the connection.
> 
> 
> Solution 1 is nearer to the old behavior. Solution 2 would mean a complete rewrite of the communication stack, since it is builded upon the idea of having steady connections and an abort is something unusual. It would be nice, if you implement keep alive messages.
> 
> TL;DR:

**Comment by ostcar** [comment]
@FinnStutzenstein 

I can not reproduce this. I opened a connection in firefox and chromium and it did not break for more then 10 minutes.

The link fro ietf.org is from 2011. Can you find any newer informations about this issue or help me to reproduce it?

The solution 2 is not possible for OpenSlides4 thins it would send all requred data every 120 Seconds. If this is really a problem, then we have to send some keep alive data. Could an empty object `{}` work?

Before we do this, it would be nice to get the information, how often it has to be  send from a good and up-to-date source.

**Comment by FinnStutzenstein** [comment]
I still have this problem. An empty object is good IMO. Do you currently have some kind of settings? The keep-alive interval should be configurable as well are the redis host, port, OS host, port. I think env variables should be sufficient.

**Comment by ostcar** [comment]
See #27

Could you ask around if it is possible to fix this issue on another network layer? Maybe it is possible on the tcp or http layer to send keep alive messages that are ignored by the application layer. Maybe there is a better solution in http2?

**Comment by FinnStutzenstein** [comment]
We need http/1.1 for development (otherwise you have to deal with https...). This is not tcp related (tcp KAs is not the problem here), and the http keep-alive header is for handling multiple requests through the same tcp connection. This is not the case here either. We have one request, and from the clients point of view, there is no response. I can understand, if browsers (and nginx and haproxy also have timeouts for this) close the connection, if there is no response.

**Comment by ostcar** [comment]
@FinnStutzenstein can this be closed?
